File: services/scala-play/app/services/PubSubService.scala
package services

import com.google.api.gax.core.{CredentialsProvider, NoCredentialsProvider}
import com.google.api.gax.grpc.GrpcTransportChannel
import com.google.api.gax.rpc.{FixedTransportChannelProvider, TransportChannelProvider}
import com.google.cloud.pubsub.v1.{Publisher, TopicAdminClient, TopicAdminSettings}
import com.google.protobuf.ByteString
import com.google.pubsub.v1.{PubsubMessage, TopicName}
import io.grpc.ManagedChannelBuilder
import play.api.libs.json.*
import play.api.{Configuration, Logging}

import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit
import javax.inject.*
import scala.jdk.CollectionConverters.*
import scala.util.Try

/** Service for publishing interactions to Google Cloud Pub/Sub. */
@Singleton
class PubSubService @Inject() (config: Configuration) extends Logging:

  private val projectId: Option[String] =
    config.getOptional[String]("gcp.projectId")
  private val topicName: Option[String] =
    config.getOptional[String]("pubsub.topic")
  private val emulatorHost: Option[String] =
    sys.env.get("PUBSUB_EMULATOR_HOST")

  private val publisher: Option[Publisher] = for
    project <- projectId
    topic <- topicName
  yield createPublisher(project, topic)

  private def createPublisher(project: String, topic: String): Publisher =
    val topicNameObj = TopicName.of(project, topic)

    val builder = Publisher.newBuilder(topicNameObj)

    emulatorHost match
      case Some(host) =>
        logger.info(s"Using Pub/Sub emulator at $host")
        // Configure for emulator (no credentials, insecure channel)
        val channel = ManagedChannelBuilder.forTarget(host).usePlaintext().build()
        val channelProvider: TransportChannelProvider =
          FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel))
        val credentialsProvider: CredentialsProvider = NoCredentialsProvider.create()

        builder
          .setChannelProvider(channelProvider)
          .setCredentialsProvider(credentialsProvider)

        // Create topic if it doesn't exist (for emulator)
        ensureTopicExists(project, topic, channelProvider, credentialsProvider)

      case None =>
        // Use default credentials for production

    logger.info(s"Pub/Sub configured: $topicNameObj")
    builder.build()

  private def ensureTopicExists(
      project: String,
      topic: String,
      channelProvider: TransportChannelProvider,
      credentialsProvider: CredentialsProvider
  ): Unit =
    Try {
      val settings = TopicAdminSettings
        .newBuilder()
        .setTransportChannelProvider(channelProvider)
        .setCredentialsProvider(credentialsProvider)
        .build()
      val adminClient = TopicAdminClient.create(settings)
      try
        val topicNameObj = TopicName.of(project, topic)
        try adminClient.getTopic(topicNameObj)
        catch
          case _: com.google.api.gax.rpc.NotFoundException =>
            adminClient.createTopic(topicNameObj)
            logger.info(s"Created topic: $topicNameObj")
      finally adminClient.close()
    }.recover { case e: Exception =>
      logger.warn(s"Failed to ensure topic exists: ${e.getMessage}")
    }

  /** Publish sanitized interaction to Pub/Sub.
    *
    * @param interaction
    *   The interaction JSON object (will be sanitized before publishing)
    */
  def publish(interaction: JsObject): Unit =
    publisher match
      case None => // Pub/Sub not configured
      case Some(pub) =>
        Try {
          val sanitized = sanitizeInteraction(interaction)
          val data = ByteString.copyFromUtf8(Json.stringify(sanitized))

          val attributes = buildAttributes(interaction)

          val message = PubsubMessage
            .newBuilder()
            .setData(data)
            .putAllAttributes(attributes.asJava)
            .build()

          val future = pub.publish(message)
          // Wait for publish to complete (with timeout)
          future.get(10, TimeUnit.SECONDS)
        }.recover { case e: Exception =>
          logger.error(s"Failed to publish to Pub/Sub: ${e.getMessage}")
        }

  /** Sanitize interaction for Pub/Sub (remove sensitive fields like token). */
  private def sanitizeInteraction(interaction: JsObject): JsObject =
    // Copy only safe fields (explicitly exclude "token")
    val safeFields = Seq(
      "type",
      "id",
      "application_id",
      "data",
      "guild_id",
      "channel_id",
      "member",
      "user",
      "locale",
      "guild_locale"
    )

    val sanitizedFields = safeFields.flatMap { field =>
      (interaction \ field).asOpt[JsValue].map(field -> _)
    }

    JsObject(sanitizedFields)

  /** Build Pub/Sub message attributes. */
  private def buildAttributes(interaction: JsObject): Map[String, String] =
    val base = Map(
      "interaction_id" -> (interaction \ "id").asOpt[String].getOrElse(""),
      "interaction_type" -> (interaction \ "type")
        .asOpt[Int]
        .map(_.toString)
        .getOrElse(""),
      "application_id" -> (interaction \ "application_id")
        .asOpt[String]
        .getOrElse(""),
      "guild_id" -> (interaction \ "guild_id").asOpt[String].getOrElse(""),
      "channel_id" -> (interaction \ "channel_id").asOpt[String].getOrElse(""),
      "timestamp" -> DateTimeFormatter.ISO_INSTANT.format(Instant.now())
    )

    // Add command name if available
    val commandName = (interaction \ "data" \ "name").asOpt[String]
    commandName.fold(base)(name => base + ("command_name" -> name))

  /** Shutdown the publisher gracefully. */
  def shutdown(): Unit =
    publisher.foreach { pub =>
      pub.shutdown()
      pub.awaitTermination(10, TimeUnit.SECONDS)
    }

File: services/java-quarkus/src/main/java/com/discord/webhook/service/PubSubService.java
package com.discord.webhook.service;

import com.discord.webhook.model.Interaction;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.api.core.ApiFuture;
import com.google.api.core.ApiFutureCallback;
import com.google.api.core.ApiFutures;
import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GrpcTransportChannel;
import com.google.api.gax.rpc.FixedTransportChannelProvider;
import com.google.cloud.pubsub.v1.Publisher;
import com.google.common.util.concurrent.MoreExecutors;
import com.google.protobuf.ByteString;
import com.google.pubsub.v1.PubsubMessage;
import com.google.pubsub.v1.TopicName;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.quarkus.runtime.Shutdown;
import io.quarkus.runtime.Startup;
import jakarta.annotation.PostConstruct;
import jakarta.enterprise.context.ApplicationScoped;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.jboss.logging.Logger;

@ApplicationScoped
@Startup
public class PubSubService {

  private static final Logger logger = Logger.getLogger(PubSubService.class);

  @ConfigProperty(name = "google.cloud.project")
  Optional<String> projectId;

  @ConfigProperty(name = "pubsub.topic")
  Optional<String> topicName;

  @ConfigProperty(name = "pubsub.emulator.host")
  Optional<String> emulatorHost;

  private Publisher publisher;
  private ManagedChannel channel;
  private final ObjectMapper objectMapper = new ObjectMapper();

  @PostConstruct
  public void init() {
    if (projectId.isEmpty()
        || topicName.isEmpty()
        || projectId.get().isEmpty()
        || topicName.get().isEmpty()) {
      logger.info("Pub/Sub not configured (missing project or topic)");
      return;
    }

    try {
      TopicName topic = TopicName.of(projectId.get(), topicName.get());
      Publisher.Builder publisherBuilder = Publisher.newBuilder(topic);

      // Configure for emulator if specified
      if (emulatorHost.isPresent() && !emulatorHost.get().isEmpty()) {
        channel = ManagedChannelBuilder.forTarget(emulatorHost.get()).usePlaintext().build();

        publisherBuilder
            .setChannelProvider(
                FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel)))
            .setCredentialsProvider(NoCredentialsProvider.create());

        logger.info("Pub/Sub configured for emulator: " + emulatorHost.get());
      }

      publisher = publisherBuilder.build();
      logger.info("Pub/Sub publisher initialized for topic: " + topicName.get());
    } catch (Exception e) {
      logger.warn("Failed to initialize Pub/Sub publisher: " + e.getMessage());
    }
  }

  @Shutdown
  public void shutdown() {
    if (publisher != null) {
      try {
        publisher.shutdown();
        publisher.awaitTermination(10, TimeUnit.SECONDS);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        logger.warn("Interrupted while shutting down publisher");
      }
    }
    if (channel != null) {
      try {
        channel.shutdown();
        channel.awaitTermination(5, TimeUnit.SECONDS);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
      }
    }
  }

  public boolean isConfigured() {
    return publisher != null;
  }

  public void publishAsync(Interaction interaction) {
    if (publisher == null) {
      return;
    }

    try {
      // Create sanitized copy (remove token)
      Interaction sanitized = interaction.createSanitizedCopy();
      String json = objectMapper.writeValueAsString(sanitized);

      // Build message with attributes
      PubsubMessage.Builder messageBuilder =
          PubsubMessage.newBuilder().setData(ByteString.copyFromUtf8(json));

      // Add attributes
      Map<String, String> attributes = messageBuilder.getMutableAttributes();
      if (interaction.getId() != null) {
        attributes.put("interaction_id", interaction.getId());
      }
      attributes.put("interaction_type", String.valueOf(interaction.getType()));
      if (interaction.getApplicationId() != null) {
        attributes.put("application_id", interaction.getApplicationId());
      }
      if (interaction.getGuildId() != null) {
        attributes.put("guild_id", interaction.getGuildId());
      }
      if (interaction.getChannelId() != null) {
        attributes.put("channel_id", interaction.getChannelId());
      }
      attributes.put("timestamp", DateTimeFormatter.ISO_INSTANT.format(Instant.now()));

      // Add command name if available
      if (interaction.getData() != null && interaction.getData().containsKey("name")) {
        Object name = interaction.getData().get("name");
        if (name instanceof String) {
          attributes.put("command_name", (String) name);
        }
      }

      PubsubMessage message = messageBuilder.build();
      ApiFuture<String> future = publisher.publish(message);

      ApiFutures.addCallback(
          future,
          new ApiFutureCallback<String>() {
            @Override
            public void onSuccess(String messageId) {
              logger.debug("Published message: " + messageId);
            }

            @Override
            public void onFailure(Throwable t) {
              logger.error("Failed to publish message: " + t.getMessage());
            }
          },
          MoreExecutors.directExecutor());

    } catch (JsonProcessingException e) {
      logger.error("Failed to serialize interaction: " + e.getMessage());
    }
  }
}
